web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts

```diff
@@ -0,0 +1,210 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { z } from 'zod';
+
+import { Platform } from 'teleterm/mainProcess/types';
+
+export function invalidKeyCodeIssue(wrongKeyCode: string): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `"${wrongKeyCode}" cannot be used as a key code.`,
+  };
+}
+
+export function invalidModifierIssue(wrongModifiers: string[]): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `${wrongModifiers
+      .map(m => `"${m}"`)
+      .join(', ')} cannot be used as a modifier.`,
+  };
+}
+
+export function duplicateModifierIssue(): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `Duplicate modifier found.`,
+  };
+}
+
+export function missingModifierIssue(keyCode: string): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `"${keyCode}" requires a modifier.`,
+  };
+}
+
+export function getKeyboardShortcutSchema(platform: Platform) {
+  const allowedModifiers = getSupportedModifiers(platform);
+
+  return z
+    .string()
+    .transform(s => s.trim().split(/\s?\+\s?/))
+    .transform(putModifiersFirst(allowedModifiers))
+    .superRefine(validateKeyCodeAndModifiers(allowedModifiers))
+    .transform(adjustCasing)
+    .transform(s => s.join('+'));
+}
+
+function putModifiersFirst(
+  allowedModifiers: string[]
+): (tokens: string[]) => string[] {
+  return tokens =>
+    tokens.sort((a, b) => {
+      if (allowedModifiers.indexOf(a) === -1) {
+        return 1;
+      }
+      if (allowedModifiers.indexOf(b) === -1) {
+        return -1;
+      }
+      return allowedModifiers.indexOf(a) - allowedModifiers.indexOf(b);
+    });
+}
+
+/** Currently, works only for single characters.
+ * An accelerator string is compared to a string generated from event properties
+ * (so it should be case-insensitive too).
+ * But what is more important, this string is used also for tooltips.
+ * If we allow "CTRL+SHIFT+PAGEUP" then we cannot display it as "Ctrl+Shift+PageUp".
+ */
+function adjustCasing(tokens: string[]): string[] {
+  return tokens.map(token => {
+    if (token.length === 1) {
+      return token.toUpperCase();
+    }
+    return token;
+  });
+}
+
+function validateKeyCodeAndModifiers(
+  allowedModifiers: string[]
+): (tokens: string[], ctx: z.RefinementCtx) => void {
+  return (tokens, ctx) => {
+    // empty accelerator disables the shortcut
+    if (tokens.join('') === '') {
+      return z.NEVER;
```

Comment:
I don't think we need to return early from any of the current validations. This way if I do something like `"Command+Command+ł"` I'm going to see all errors immediately without having to fix them one by one (and restarting the app each time I attempt to fix an error).

Reply:
Done, but keep in mind that issues are not grouped, so every issue is rendered as a separate `<li>` (but I think it is ok, we don't have enough space to render more sophisticated structures there).